handshake/crypto_setup_server.go

```diff
@@ -152,6 +153,9 @@ func (h *cryptoSetupServer) handleMessage(chloData []byte, cryptoData map[Tag][]
 		if err != nil {
 			return false, err
 		}
+		h.mutex.Lock()
+		h.wroteSHLO = true
```

Comment:
I don't see how to fix this without putting the `Write` into the mutex as well. And this will deadlock for sure.